torch/csrc/jit/script/slot.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+namespace jit {
+namespace script {
+
+// a stable location that can hold an IValue.
+// Currently this is internally implemented as a pointer, but when
+// modules become first-class this will be a pair of  <module_ivalue, slot_number>
+struct Slot {
+  friend struct NamedIValue;
+  Slot()
+  : slot_(nullptr) {}
+  Slot(at::IValue* slot)
+  : slot_(slot) {}
+  at::IValue& operator*() const {
+    return *slot_;
+  }
+  at::IValue* operator->() const {
+      return slot_;
+  }
+  bool operator==(const Slot& rhs) const {
+    return slot_ == rhs.slot_;
+  }
+private:
+  at::IValue* slot_;
+  friend struct std::hash<Slot>;
```

Comment:
I don't know if `return std::hash<at::IValue*>{}(&(*s.slot_));` is any better than declaring a friend.

Reply:
<in person>: This particular implementation will only live until a later diff in this stack merges Slot and NamedIValue, so I will keep it as is for now.